File: 0366-find-leaves-of-binary-tree/0366-find-leaves-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findLeaves(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = defaultdict(list)
        
        def dfs(root):
            if not root:
                return 0
                
            left = dfs(root.left)
            right = dfs(root.right)
            
            cur_l = 1 + max(left, right)
            res[cur_l].append(root.val)
            
            return cur_l
        
        dfs(root)
        return res.values()
                    

File: 0366-find-leaves-of-binary-tree/0366-find-leaves-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<List<Integer>> res;
    
    private int getHeight(TreeNode root) {
        
        if (root == null) {
            return -1;
        }
        
        int leftHeight = getHeight(root.left);
        int rightHeight = getHeight(root.right);
        
        int currHeight = 1 + Math.max(leftHeight, rightHeight);
        
        if (this.res.size() == currHeight) {
            this.res.add(new ArrayList<>());
        }
        
        this.res.get(currHeight).add(root.val);
        
        return currHeight;
    }
    
    public List<List<Integer>> findLeaves(TreeNode root) {
        this.res = new ArrayList<>();
        
        getHeight(root);
        
        return this.res;
    }
}